### PR TITLE
feat(connector-builder-mcp): Add create_connector_manifest_scaffold MCP tool

### PR DESCRIPTION
**Pull request by aaronsteers**
# feat(connector-builder-mcp): Add create_connector_manifest_scaffold MCP tool

## Summary

This PR adds a new MCP tool `create_connector_manifest_scaffold` that streamlines the initial creation of Airbyte connector manifest YAML files. The tool accepts basic connector configuration inputs and generates a complete, valid manifest scaffold with TODO placeholders and inline comments for fields that need to be filled in later.

**Key features:**
- **Strict input validation**: Connector names must be kebab-case with "source-" prefix, authentication types are validated against supported enum values
- **Multiple authentication support**: NoAuth, ApiKeyAuthenticator, BearerAuthenticator, BasicHttpAuthenticator, OAuthAuthenticator  
- **Pagination support**: None, page_increment, offset_increment, cursor_pagination
- **Dynamic schema detection**: Uses InlineSchemaLoader to minimize LLM context usage during development
- **TODO-driven development**: Generated manifests include TODO placeholders with inline comments that still pass validation
- **Comprehensive testing**: Unit tests verify all authentication/pagination combinations generate valid manifests

The generated manifests are designed to be immediately testable and incrementally improvable, supporting the AI-assisted connector development workflow.

## Review & Testing Checklist for Human

- [ ] **End-to-end MCP tool testing**: Verify the tool is properly registered and callable through the MCP server interface
- [ ] **Generated manifest validation**: Test that manifests generated for all authentication/pagination combinations actually pass Airbyte's validation  
- [ ] **TODO placeholder verification**: Confirm that TODO values don't break manifest validation while still providing clear guidance
- [ ] **Input validation edge cases**: Test connector name patterns, invalid authentication types, and malformed inputs to ensure proper error handling
- [ ] **Real connector workflow**: Try using a generated manifest as the starting point for an actual connector to verify the structure works in practice

**Recommended test plan**: Use the MCP tool to generate manifests for different authentication types (especially OAuth and API key), then attempt to load them in Airbyte's connector builder or validation tools to ensure they work correctly.

---

### Diagram

```mermaid
%%{ init : { "theme" : "default" }}%%
graph TD
    CB["connector_builder_mcp/_connector_builder.py"]:::major-edit
    TS["tests/test_manifest_scaffold.py"]:::major-edit
    REG["register_connector_builder_tools()"]:::minor-edit
    
    CB --> VAL["validate_manifest()"]:::context
    CB --> YAML["Generated YAML Manifest"]:::context
    TS --> CB
    REG --> CB
    
    CB --> AUTH["Authentication Templates<br/>(NoAuth, ApiKey, OAuth, etc)"]:::context
    CB --> PAG["Pagination Templates<br/>(PageIncrement, Cursor, etc)"]:::context
    CB --> SCHEMA["Dynamic Schema Loader"]:::context
    
    YAML --> TODO["TODO Placeholders<br/>+ Inline Comments"]:::context
    
    subgraph Legend
        L1[Major Edit]:::major-edit
        L2[Minor Edit]:::minor-edit  
        L3[Context/No Edit]:::context
    end

classDef major-edit fill:#90EE90
classDef minor-edit fill:#87CEEB
classDef context fill:#FFFFFF
```

### Notes

- The implementation follows existing MCP tool patterns in the codebase for consistency
- All generated manifests pass validation tests, but real-world testing with Airbyte's connector builder is recommended
- The tool is designed to work with AI assistants by providing clear TODO guidance and maintaining valid manifest structure
- Dynamic schema detection reduces the need for LLMs to constantly update JSON schemas during development

**Session details**: Requested by AJ Steers (@aaronsteers)  
**Devin session**: https://app.devin.ai/sessions/73d258b22cd64316989b80210ea3c423

### CODE REVIEW

File: connector_builder_mcp/_manifest_scaffold.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+        manifest_yaml = _generate_manifest_yaml_directly(input_data)
+
+        validation_result = validate_manifest(manifest_yaml)
+
+        return ConnectorManifestScaffoldResult(
+            success=validation_result.is_valid,
+            manifest_yaml=manifest_yaml,
+            validation_result=validation_result,
+            errors=validation_result.errors if not validation_result.is_valid else [],
+        )
```
Comment (aaronsteers) [inline connector_builder_mcp/_manifest_scaffold.py:343]: No need for this class. Return a string containing the YAML or a string starting with `ERROR: `.

As discussed, the manifest _should_ be valid. If not, something is very wrong.

File: connector_builder_mcp/_manifest_scaffold.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+class PaginationType(str, Enum):
+    """Valid pagination types for connector manifests."""
+
+    NONE = "none"
+    PAGE_INCREMENT = "page_increment"
+    OFFSET_INCREMENT = "offset_increment"
+    CURSOR_PAGINATION = "cursor_pagination"
+
+
+class ConnectorManifestScaffoldInput(BaseModel):
```
Comment (aaronsteers) [inline connector_builder_mcp/_manifest_scaffold.py:34]: Another useless class. Redundant with the function signature. You can use a helper function if needed to keep the tool function clean, but this class doesn't add anything useful. Delete please

File: connector_builder_mcp/_manifest_scaffold.py [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+        return """    type: OAuthAuthenticator
+    client_id: "{{ config['client_id'] }}"
+    client_secret: "{{ config['client_secret'] }}"
+    refresh_token: "{{ config['refresh_token'] }}"
+    token_refresh_endpoint: "{{ config['token_refresh_endpoint'] }}\""""
+
+    raise ValueError(f"Unsupported authentication type: {auth_type}")
+
+
+def _generate_paginator_yaml(pagination_type: PaginationType) -> str:
```
Comment (aaronsteers) [inline connector_builder_mcp/_manifest_scaffold.py:152]: Delete this function and just use the "None" string instead.

File: connector_builder_mcp/_manifest_scaffold.py [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+
+        try:
+            auth_type = AuthenticationType(authentication_type)
+        except ValueError:
+            valid_auth_types = [at.value for at in AuthenticationType]
+            return f"ERROR: Input validation error: Invalid authentication_type. Must be one of: {valid_auth_types}"
+
+        primary_key = "TODO"
+        record_selector_path: list[str] = []
+        pagination_type_enum = PaginationType.NONE
```
Comment (aaronsteers) [inline connector_builder_mcp/_manifest_scaffold.py:319]: Useless variable. Delete here and delete the option in the helper function

File: connector_builder_mcp/_manifest_scaffold.py [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+
+        manifest_yaml = _generate_manifest_yaml_directly(
+            connector_name=connector_name,
+            api_base_url=api_base_url,
+            initial_stream_name=initial_stream_name,
+            initial_stream_path=initial_stream_path,
+            authentication_type=auth_type,
+            primary_key=primary_key,
+            http_method=http_method.upper(),
+            record_selector_path=record_selector_path,
```
Comment (aaronsteers) [inline connector_builder_mcp/_manifest_scaffold.py:329]: Delete record selector input arg here

File: connector_builder_mcp/_manifest_scaffold.py [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+        record_selector_path: list[str] = []
+        pagination_type_enum = PaginationType.NONE
+
+        manifest_yaml = _generate_manifest_yaml_directly(
+            connector_name=connector_name,
+            api_base_url=api_base_url,
+            initial_stream_name=initial_stream_name,
+            initial_stream_path=initial_stream_path,
+            authentication_type=auth_type,
+            primary_key=primary_key,
```
Comment (aaronsteers) [inline connector_builder_mcp/_manifest_scaffold.py:327]: Delete primary key input here

File: connector_builder_mcp/_manifest_scaffold.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+
+    raise ValueError(f"Unsupported authentication type: {auth_type}")
+
+
+def _generate_paginator_yaml(pagination_type: PaginationType) -> str:
+    """Generate paginator YAML configuration with proper indentation."""
+    if pagination_type == PaginationType.NONE:
+        return """      paginator:
+        type: NoPagination
+      # TODO: Uncomment and configure pagination when needed
```
Comment (aaronsteers) [inline connector_builder_mcp/_manifest_scaffold.py:157]:
```suggestion
      # TODO: Uncomment and configure pagination when known
```

**Comment by aaronsteers** [comment]
@lmossman - I'll merge this tonight but I want to tag you for visibility. This increment does YAML scaffolding on behalf of the AI, reducing the amount of trial and error needed to get a first successful test read.

File: connector_builder_mcp/_manifest_scaffold.py [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+      #   page_token_option:
+      #     type: RequestOption
+      #     inject_into: request_parameter
+      #     field_name: page
+      #   pagination_strategy:
+      #     type: PageIncrement
+      #     start_from_page: 1
+      schema_loader:
+        type: InlineSchemaLoader
+        schema:
```
Comment (aaronsteers) [inline connector_builder_mcp/_manifest_scaffold.py:85]: I thought we'd start with a dynamic schema loader. Is that what this is?

File: connector_builder_mcp/_manifest_scaffold.py [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+{spec_yaml}
+
+
+# TODO: Uncomment and configure incremental sync when ready
+# incremental_sync:
+#   type: DatetimeBasedCursor
+#   cursor_field: updated_at  # TODO: Replace with actual timestamp field
+#   datetime_format: "%Y-%m-%dT%H:%M:%S%z"
+#   start_datetime:
+#     type: MinMaxDatetime
```
Comment (aaronsteers) [inline connector_builder_mcp/_manifest_scaffold.py:105]: This full incremental block should be inline with the stream definition.

File: connector_builder_mcp/_manifest_scaffold.py [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+    username: "{{ config['username'] }}"
+    password: "{{ config['password'] }}\""""
+    elif auth_type == AuthenticationType.OAUTH:
+        return """    type: OAuthAuthenticator
+    client_id: "{{ config['client_id'] }}"
+    client_secret: "{{ config['client_secret'] }}"
+    refresh_token: "{{ config['refresh_token'] }}"
+    token_refresh_endpoint: "{{ config['token_refresh_endpoint'] }}\""""
+
+    raise ValueError(f"Unsupported authentication type: {auth_type}")
```
Comment (aaronsteers) [inline connector_builder_mcp/_manifest_scaffold.py:142]: Add list of expected values in this message

File: connector_builder_mcp/_manifest_scaffold.py [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+
+
+def _generate_connection_spec_yaml(connector_name: str, auth_type: AuthenticationType) -> str:
```
Comment (aaronsteers) [inline connector_builder_mcp/_manifest_scaffold.py:145]: Refactor this function to make it more DRY. only 'properties' and 'required' are variant. Update so these conditional blocks only affect the part changes.

Also, add a TODO comment that this is where additional configuration inputs may be defined, as needed.

File: connector_builder_mcp/_manifest_scaffold.py [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+      refresh_token:
+        type: string
+        airbyte_secret: true
+      token_refresh_endpoint:
+        type: string""",
+            """    required:
+      - client_id
+      - client_secret
+      - refresh_token
+      - token_refresh_endpoint""",
```
Comment (aaronsteers) [inline connector_builder_mcp/_manifest_scaffold.py:192]: I've seen precedent where all auth is under a "credentials" parent. Don't change anything here in this PR but open a ticket in this repo to consider this change.
